[HUDI-6211] Fix reading of schema-evolved complex columns for Flink

voonhous opened the pull request:
### Change Logs

If a Hudi-table's column with a complex type (struct) is evolved with newly added fields, or reordered fields, Hudi-On-Flink will not be able to read the parquet files that are in the old schema.

An error as such will be thrown:

Different errors will be thrown for schema evolution operations. If a new column is added at the back of the struct column, the error below will be thrown.

```log
Caused by: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.get(ArrayList.java:435)
	at org.apache.parquet.schema.GroupType.getType(GroupType.java:216)
	at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.createWritableColumnVector(ParquetSplitReaderUtil.java:528)
	at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.createWritableColumnVector(ParquetSplitReaderUtil.java:416)
	at org.apache.hudi.table.format.cow.vector.reader.ParquetColumnarRowSplitReader.createWritableVectors(ParquetColumnarRowSplitReader.java:217)
	at org.apache.hudi.table.format.cow.vector.reader.ParquetColumnarRowSplitReader.<init>(ParquetColumnarRowSplitReader.java:157)
	at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.genPartColumnarRowReader(ParquetSplitReaderUtil.java:148)
	at org.apache.hudi.table.format.RecordIterators.getParquetRecordIterator(RecordIterators.java:72)
	at org.apache.hudi.table.format.cow.CopyOnWriteInputFormat.open(CopyOnWriteInputFormat.java:132)
	at org.apache.hudi.table.format.cow.CopyOnWriteInputFormat.open(CopyOnWriteInputFormat.java:66)
	at org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction.run(InputFormatSourceFunction.java:84)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:332)
```

If a column is added in the middle of the struct column, a error as such will be thrown:
```log
java.lang.IllegalArgumentException: Unexpected type: INT32 at org.apache.parquet.Preconditions.checkArgument(Preconditions.java:77) at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.createWritableColumnVector(ParquetSplitReaderUtil.java:456) at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.createWritableColumnVector(ParquetSplitReaderUtil.java:413) at org.apache.hudi.table.format.cow.vector.reader.ParquetColumnarRowSplitReader.createWritableVectors(ParquetColumnarRowSplitReader.java:216) at org.apache.hudi.table.format.cow.vector.reader.ParquetColumnarRowSplitReader.<init>(ParquetColumnarRowSplitReader.java:156) at org.apache.hudi.table.format.cow.ParquetSplitReaderUtil.genPartColumnarRowReader(ParquetSplitReaderUtil.java:151) at org.apache.hudi.table.format.cow.CopyOnWriteInputFormat.open(CopyOnWriteInputFormat.java:151) at org.apache.hudi.table.format.cow.CopyOnWriteInputFormat.open(CopyOnWriteInputFormat.java:68) at org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction.run(InputFormatSourceFunction.java:116) at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:128) at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:73) at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:333)
```

### Impact

_Describe any public API or user-facing feature change or any performance impact._
None

### Risk level (write none, low medium or high below)

_If medium or high, explain what verification was done to mitigate the risks._
None, adding additional logic to handle schema evolution operations on complex STRUCT/ROW types. 

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented:
Thanks for the contribution, I have reviewed and applied a patch for flink 1.17.x:
[6211.patch.zip](https://github.com/apache/hudi/files/11476712/6211.patch.zip)

Still needs to strength the test cases in following scenarios:
1. add schema evolution like adding new columns in the midst like you mentioned
2. add test cases where all the fields in the struct type are NULL.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/table/ITTestSchemaEvolution.java:
```diff
@@ -184,17 +185,18 @@ private void writeTableWithSchema1(TableOptions tableOptions) throws ExecutionEx
         + "  cast(age as int),"
         + "  cast(ts as timestamp),"
+        + "  cast(struct_col as row<f0 int, f1 string>),"
         + "  cast(`partition` as string) "
```
`struct_col` -> `f_struct`, whatever, it's just a preference.

voonhous replied:
Done!

voonhous commented:
> 2\. add

Okay!

danny0405 commented on hudi-flink-datasource/hudi-flink1.13.x/src/main/java/org/apache/hudi/table/format/cow/vector/reader/RowColumnReader.java:
```diff
@@ -44,7 +44,12 @@ public void readToVector(int readNumber, WritableColumnVector vector) throws IOE
+      // schema evolution: read the file with a new extended field name.
+      if (fieldReaders.get(i) == null) {
+        vectors[i].fillWithNulls();
+      } else {
```
Guess the vector should never be null now?

danny0405 commented on hudi-flink-datasource/hudi-flink1.13.x/src/main/java/org/apache/hudi/table/format/cow/ParquetSplitReaderUtil.java:
```diff
@@ -387,14 +387,20 @@ private static ColumnReader createColumnReader(
+          int fieldIndex = getFieldIndexInPhysicalType(rowType.getFields().get(i).getName(), groupType);
+          if (fieldIndex < 0) {
+            fieldReaders.add(null);
+          } else {
```
Can we use mock constant reader for `NULL` constant to avoid the null reader instead? It is more safe for programming convention.

voonhous replied:
Agree, i have implemented a new ColumnReader called `EmptyColumnReader`.

voonhous commented:
@danny0405 I have applied the changes from the patch as requested.

Have also addressed the comments and added more STRUCT schema evolution scenarios in the test.

Can you please take a look at it again?

Another thing to point out, I modified the test to use a ExecutorService that is able to timeout the closeable iterator. 

I found that the original implementation of `#checkAnswer()` is not exactly correct. Actual answer is dependent on the size of Expected answer, which can introduce errors.

Example:

```txt
Correct answer = 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10

provided expected = 0, 1, 2, 3, 4, 5, 6, 7, 8
actual result = 0, 1, 2, 3, 4, 5,  6, 7, 8, 11, 12
```

It will tell you that that the test is correct.

If the **actual** size is dependent on **provided expected**, actual result might get truncated and we our test cases will pass, while in actuality, it should fail.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/table/ITTestSchemaEvolution.java:
```diff
@@ -303,22 +316,43 @@ private void checkAnswerWithMeta(TableOptions tableOptions, String... expectedRe
       }
+    };
+
+    ExecutorService executor = Executors.newSingleThreadExecutor();
```
There is no need to use async await, just iterate through the table result to fetch all the results.

voonhous replied:
https://github.com/apache/hudi/pull/8711/files/a9b183c71ec4b96012294afd42bca677fa614a82#diff-367630f5c7c72af82c5a293387290c4670546b0206dbc84a03b802cada31dc82R90

Erm, this test is a streaming read, the iterator will keep blocking. It will not exit...

danny0405 reviewed (approved):
+1, thanks for the contribution @voonhous

voonhous commented:
@danny0405 Apologies for the inconvenience, made another force push to fix checkstyle errors:

```log
 src/main/java/org/apache/hudi/table/format/cow/vector/reader/EmptyColumnReader.java:[38] (whitespace) EmptyLineSeparator: 'METHOD_DEF' has more than 1 empty lines before.
```

danny0405 commented:
> @danny0405 Apologies for any inconvenience caused, made another force push to fix checkstyle errors:
> 
> ```
>  src/main/java/org/apache/hudi/table/format/cow/vector/reader/EmptyColumnReader.java:[38] (whitespace) EmptyLineSeparator: 'METHOD_DEF' has more than 1 empty lines before.
> ```

You may need to install the `hudi-flink` module in you local and make it pass.

voonhous commented:
> > @danny0405 Apologies for any inconvenience caused, made another force push to fix checkstyle errors:
> > ```
> >  src/main/java/org/apache/hudi/table/format/cow/vector/reader/EmptyColumnReader.java:[38] (whitespace) EmptyLineSeparator: 'METHOD_DEF' has more than 1 empty lines before.
> > ```
> 
> You may need to install the `hudi-flink` module in you local and make it pass.

Done! Installed it locally and fixed all the checkstyle violations. Fingers crossed that there are no more checkstyle violations.

voonhous commented:
@danny0405 CI is finally green, thank you for helping to review it!
